### PR TITLE
Block validation methods, take 3

### LINKED ISSUE #1879: Block: Low-Level Validation Method Refactor
This is a (relatively) modest proposal to refactor the Block validation methods and logic as one of the last bigger tasks on the way to the breaking releases. Major outcome would be that we would free ourself from the problematic (and in some sense circular) `Blockchain` dependency and that subclassing of `Block` or `Header` for consensus specific customizations would get easier.

Lots of little tasks which will sum up to - hopefully - a consistent set of improvements:

### BlockHeader

- [ ] `_validateHeaderFields()` called in the constructor should be renamed to `_genericFormatValidation()`
- [ ] `Header.validate()`:
  - [ ] The first upper consensus-specific checks are misplaced here since these are just static format checks as well. These should be moved to a separate `_consensusFormatValidation()` method called from the constructor after `_genericFormatValidation()`. This also allows for easier consensus-specific replacement on subclassing
  - [ ] The rest of the `validate()` functionality (Including the following consensus-specific checks) should be moved to a new method in the `Blockchain` class `validateBlockHeader(header: BlockHeader, height)`
- [ ] The `validateDifficulty()` and `validateCliqueDifficulty()` methods should be unified to `validateDifficulty()` with appropriate consensus checks
  - [ ] Subtask: same for the `Block` class

### Block

- [ ] `validateUncles()`: method should be moved to `Blockchain` as `validateBlockUncles(block: Block)`
  - [ ] Same with private `_validateUncleHeaders()` method called from within the above method
- [ ] `validate()`: method can be removed without replacement
- [ ] Bonus task: The `onlyHeader` param on the `validateData()` method is a bit quirky, wonder if this can be removed along the refactoring (or in general)?
- [ ] TBD - also on th `validateData()` method: this method is a bit the corresponding method to `_genericFormatValidation()` in `BlockHeader` if I am not mistaken. I wonder if we should align here and a) remove this method (respectively rename to private `_genericFormatValidation()`) and then call from the constructor? 🤔

### Generic

- [ ] Remove `@ethereumjs/blockchain` dependency (good check if nothing gots forgotten)

I think @emersonmacro would be a good candidate for this task, since he already dealt with the `Blockchain` refactoring and this is going somewhat along the same line. 🙂 Haven't got a confirmation from him though yet if he is available and think this fits.

### PR DESCRIPTION
**Pull request by acolytec3**
Fixes #1879 and #1889

- Removes `block.validate` and `blockheader.validate` and moves the format/basic consensus checks into helper methods that are called in the respective constructors.  These helpers throw if invalid fields are found
- Creates two new methods in `blockchain`, `validateBlock` and `validateHeader` that are called when a block is inserted into the chain.  All of the checks from the removed `validate` methods that require a blockchain context are now done in these methods.
- Move `validateUncles` to `blockchain
- Rewrite block tests to remove `blockchain` dependency where possible and moved to `blockchain` where not
- Some test data was modified (notably modifying the extradata fields to be PoA compliant) to make it work with the tests
- fixes qheap import issue related to #1975

### CODE REVIEW

**Comment by acolytec3** [comment]
@holgerd77 here's my WIP on the block validation methods.  `block.spec.ts` tests don't pass at present since I need to address the changes made with moving the header validation methods around but I've taken the first step of moving header validation to blockchain and am going to focus on the difficulty checks next.  Blockchain tests are all currently passing so hopefully won't have to adjust them too much more once I finish the difficulty check updates.

**Comment by acolytec3** [comment]
Correction: all the blockchain tests are passing locally.  CI is failing because I haven't addressed the broader monorepo yet and was focused on getting things working between `block` and `blockchain` and would then address other places where build/tests are failing.

File: packages/block/src/block.ts
Comment (holgerd77) [inline packages/block/src/block.ts:281]: I think I wrote otherwise in the original issue but I guess we really want to keep this combined version (which would also solve a related client-TODO) in a public `Blockchain.validateBlock()` method.

Comment (acolytec3) [reply packages/block/src/block.ts:281]: The thought being we just want to be able to validate a new block in it's entirety without having to actually put it in the chain?

Comment (holgerd77) [reply packages/block/src/block.ts:281]: (or even better in the cases where this is possible: both doesn't need to be in the chain, so these would be the cases where a method could even be called in a static context)

Comment (acolytec3) [reply packages/block/src/block.ts:281]: I've added a public `blockchain.validateBlock` method that wraps up those three validation checks.  I don't think we can make it a static method without breaking it into further pieces since some the checks are heavily dependent on the existence of the chain.

**Comment by acolytec3** [comment]
@holgerd77 This PR is now ready for review.  I've revised the initial description to reflect the changes made.

Thanks @g11tech for helping track down some of the trickier client test failures!

**Review by g11tech — APPROVED** [review]
LGTM!

File: packages/block/src/header.ts
Comment (holgerd77) [inline packages/block/src/header.ts:526]: Can you put this method upper in the method order and place it directly under `_genericFormatValidation()`? Looks a bit misplaced here.

File: packages/block/test/difficulty.spec.ts
Comment (holgerd77) [inline packages/block/test/difficulty.spec.ts:13]: Argh, I am not allow to comment on the `canonicalDifficulty()` method itself. 😜

So here: I would suggest that we rename this to `ethashCanonicalDifficulty()` to align with the clique-specific naming and also place this method further down the line directly above the `clique*()` methods to group these consensus-specific functions together.

File: packages/block/src/block.ts
Comment (holgerd77) [inline packages/block/src/block.ts:280]: Again, can't comment on the method itself: I would have a tendency to at least rename this to `validateBlockData()`, I always stumble upon this name since data is so ambiguous. If you go along maybe just do.

File: packages/blockchain/src/consensus/casper.ts
Comment (holgerd77) [inline packages/blockchain/src/consensus/casper.ts:18]: Can we have some new lines between all the methods here (actually also the empty ones)? This looks pretty crambled.

File: packages/blockchain/src/consensus/clique.ts
Comment (holgerd77) [inline packages/blockchain/src/consensus/clique.ts:38]: Can we change the order of constants and types here (or is it just me who feel that constants would belong to the top?) and then also put all *exported* constants grouped together first?

File: packages/blockchain/src/index.ts
Comment (holgerd77) [inline packages/blockchain/src/index.ts:699]: I guess we can (and should) remove this PoW check here, to make sure this works in a generic context.

File: packages/blockchain/src/index.ts
Comment (holgerd77) [inline packages/blockchain/src/index.ts:695]: Now we have this *one* consensus-related check back in here which is unlucky. Wonder if there is a good place in `consensus` we can safely move.

Maybe in the `clique.validate()` method as well? 🤔 Might need a double check if the semantic result would be somewhat matching/equivalent.

But we should really target to remove here.

**Review by holgerd77 — CHANGES_REQUESTED** [review]
Thanks, looks already pretty great, made some comments and change requests on things like structure, (method) order, naming and visibility of things.

File: packages/block/src/header.ts
Comment (acolytec3) [reply packages/block/src/header.ts]: Nope, fixed.

File: packages/blockchain/src/consensus/clique.ts
Comment (acolytec3) [reply packages/blockchain/src/consensus/clique.ts]: Done

File: packages/blockchain/src/consensus/clique.ts
Comment (acolytec3) [reply packages/blockchain/src/consensus/clique.ts]: Thanks!  Didn't notice that before

File: packages/block/src/block.ts
Comment (acolytec3) [reply packages/block/src/block.ts:280]: Done

File: packages/blockchain/src/index.ts
Comment (acolytec3) [reply packages/blockchain/src/index.ts:692]: This one is super tricky I've discovered.  Because `this.consensus` is the blockchain's current consensus mechanism, this fails on merge transition blocks where the header is now PoS but the blockchain is still PoW (since the blockchain hasn't switched to PoS so `blockchain.consensus` is `ethashConsensus`.  We should skip this check on the merge block because the PoS difficulty check (i.e. PoS blocks must always have difficulty 0) is done in the block constructor via `block._consensusValidation) and it will always fail if you pass in a new header that's a PoS block but the current chain is either still PoA or PoW.  I'll check for PoS instead and skip if so.

**Comment by acolytec3** [comment]
@holgerd77 I think I've addressed all the feedback at this point.

Two open questions for you/others to consider:

- Decide whether `block.validateTransactions` should throw or not (seems like it shouldn't since it can return an array of errors if multiple transactions fail
- Decide if we should add a flag to the block/blockchain constructors to allow the validation checks to be bypassed (like the `validateBlocks` parameter on the blockchain constructor

File: packages/block/src/header.ts
Comment (gabrocheleau) [inline packages/block/src/header.ts:353]: Since we know that this.number is defined, perhaps we could simplify this a bit:
```suggestion
      const isInitialEIP1559Block = this.number === this._common.hardforkBlock(Hardfork.London)
```

File: packages/blockchain/src/index.ts
Comment (gabrocheleau) [inline packages/blockchain/src/index.ts:710]:
```suggestion
      const isInitialEIP1559Block = number === this._common.hardforkBlock(Hardfork.London)
```
same as above

File: packages/blockchain/src/index.ts
Comment (gabrocheleau) [inline packages/blockchain/src/index.ts:744]:
```suggestion
    if (uncleHeaders.length === 0) {
```

File: packages/blockchain/src/index.ts
Comment (gabrocheleau) [inline packages/blockchain/src/index.ts:1363]:
```suggestion
    if (common.consensusType() === ConsensusType.ProofOfAuthority) {
```

File: packages/vm/examples/run-solidity-contract.ts
Comment (gabrocheleau) [inline packages/vm/examples/run-solidity-contract.ts:17]:
```suggestion
const block = Block.fromBlockData({ header: { extraData: Buffer.alloc(97) } }, { common })
```

File: packages/vm/tests/api/runTx.spec.ts
Comment (gabrocheleau) [inline packages/vm/tests/api/runTx.spec.ts:42]:
```suggestion
      if (vm._common.consensusType() === ConsensusType.ProofOfAuthority) {
```

**Review by gabrocheleau — APPROVED** [review]
Left some minor comments, look good to me overall.

**Comment by acolytec3** [comment]
> Left some minor comments, look good to me overall.

Thanks for the detailed review!  Will address all this cleanup.

**Comment by acolytec3** [comment]
Please don't merge this.  I've got to fix the client tests, again.

**Comment by holgerd77** [comment]
@acolytec3 thanks a lot for addressing all the feedback, this looks really great! 👍 

> Decide whether block.validateTransactions 

Yes, I guess we should not throw here.

> Decide if we should add a flag to the block/blockchain constructors to allow the validation checks to be bypassed (like the validateBlocks parameter on the blockchain constructor

Maybe I am missing something but there is a `validateBlocks` parameter for `Blockchain`? 🤔

There is also a `validateConsensus` flag, looking at that part I realized that we should rename the `consensus.validataBlockData()` methods to `consensus.validateConsensus()`, this makes things a lot more clear.

**Comment by acolytec3** [comment]
> 
> Maybe I am missing something but there is a `validateBlocks` parameter for `Blockchain`? thinking
> 
> There is also a `validateConsensus` flag, looking at that part I realized that we should rename the `consensus.validataBlockData()` methods to `consensus.validateConsensus()`, this makes things a lot more clear.

The reason I ask the question is we have format level consensus checks in the blockheader constructor that require things like PoA blocks have to have minimum 97 bytes of `extraData` which in turn means we have to populate these fields when we're instantiating blocks.  Do we want to have a `validateConsensus` flag as part of the `blockOpts` so these checks can be bypassed the way we support it at the `blockchain` level.  At the moment, these checks are enforced unless you stub them out in the test itself (with occasionally frustrating results trying to triage why a test is breaking)

**Comment by holgerd77** [comment]
> The reason I ask the question is we have format level consensus checks in the blockheader constructor that require things like PoA blocks have to have minimum 97 bytes of `extraData` which in turn means we have to populate these fields when we're instantiating blocks. Do we want to have a `validateConsensus` flag as part of the `blockOpts` so these checks can be bypassed the way we support it at the `blockchain` level. At the moment, these checks are enforced unless you stub them out in the test itself (with occasionally frustrating results trying to triage why a test is breaking)

Yes, I guess we want but can you leave it out here and instead open a short issue, since this is not breaking and we can add later (if I am not mistaken).

I will merge in #1974 once CI passes. Can you prioritize on bringing this branch up to date then so that we can merge in here as well? I would want to have the big PRs out of the way.

ugh. 🙂

Thanks (I am getting a bit stressed 😋).

File: packages/block/src/header.ts
Comment (holgerd77) [inline packages/block/src/header.ts:868]: Small nit here, this should remain `protected` to allow for sub-classing.

File: packages/block/src/block.ts
Comment (holgerd77) [inline packages/block/src/block.ts:392]: Same here.

File: packages/block/src/header.ts
Comment (acolytec3) [reply packages/block/src/header.ts:868]: Got it.  Adding this to #1982

**Pull request merged**
